### PR TITLE
Inject shard name in commit-phase multi-shard errors

### PR DESCRIPTION
**Pull request by maxenglander**
## Description

In MULTI transactional mode, errors can occur at different phases:

 1. When the transaction is started (e.g. with `BEGIN`).
 2. Inside the transaction.
 3. When the transaction is committed (e.g. with `COMMIT`).

Error messages generated within-transaction include the shard name,
e.g.:

> ERROR 1317 (70100): target: keyspace2.c0-.primary: vttablet: ...

However, messages generated in the commit phase do not:

> ERROR 1317 (70100): vttablet: rpc error: code = Canceled desc...

This commit injects shard name into commit-phase errors.

## Related Issue(s)

#10692 

## Checklist

-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by Anders-PlanetScale** [comment]
I've tested locally and this resolves the issue logged here https://github.com/planetscale/vitess-private/issues/811

File: go/vt/vtgate/tabletgateway.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ func NewTabletGateway(ctx context.Context, hc discovery.HealthCheck, serv srvtop
 		statusAggregators: make(map[string]*TabletStatusAggregator),
 	}
 	gw.setupBuffering(ctx)
-	gw.QueryService = queryservice.Wrap(nil, gw.withRetry)
+	gw.QueryService = queryservice.Wrap(queryservice.Wrap(nil, gw.withRetry), gw.withShardError)
```
Comment (deepthi) [inline go/vt/vtgate/tabletgateway.go:102]: @vitessio/query-serving when you review this..
1. What are the implications of wrapping this twice?
2. If you think that is acceptable, we probably need to benchmark this PR.

File: go/vt/vtgate/tabletgateway.go [hunk 2]
```diff
@@ -334,6 +335,13 @@ func (gw *TabletGateway) withRetry(ctx context.Context, target *querypb.Target,
 		}
 		break
 	}
+	return err
```
Comment (harshit-gangal) [inline go/vt/vtgate/tabletgateway.go:338]: we should keep `NewShardError(err, target)`

File: go/vt/vtgate/tabletgateway.go [hunk 3]
```diff
@@ -160,7 +160,8 @@ func (gw *TabletGateway) setupBuffering(ctx context.Context) {
 
 // QueryServiceByAlias satisfies the Gateway interface
 func (gw *TabletGateway) QueryServiceByAlias(alias *topodatapb.TabletAlias, target *querypb.Target) (queryservice.QueryService, error) {
-	return gw.hc.TabletConnection(alias, target)
+	qs, err := gw.hc.TabletConnection(alias, target)
+	return queryservice.Wrap(qs, gw.withShardError), NewShardError(err, target)
```
Comment (harshit-gangal) [inline go/vt/vtgate/tabletgateway.go:164]: :+1:

**Comment by maxenglander** [comment]
Thanks for comments @harshit-gangal have made your suggested changes.

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**